virtualizarr/zarr.py

```diff
@@ -277,21 +345,71 @@ def metadata_from_zarr_json(filepath: Path) -> tuple[ZArray, list[str], dict]:
     dim_names = metadata.pop("dimension_names")
 
     chunk_shape = metadata["chunk_grid"]["configuration"]["chunk_shape"]
+    shape = metadata["shape"]
+    zarr_format = metadata["zarr_format"]
 
     if metadata["fill_value"] is None:
-        fill_value = np.nan
+        raise ValueError(
+            "fill_value must be specified https://zarr-specs.readthedocs.io/en/latest/v3/core/v3.0.html#fill-value"
+        )
     else:
         fill_value = metadata["fill_value"]
 
+    all_codecs = [
+        codec
+        for codec in metadata["codecs"]
+        if codec["name"] not in ("transpose", "bytes")
+    ]
+    compressor, *filters = [
+        _configurable_to_num_codec_config(_filter) for _filter in all_codecs
+    ]
     zarray = ZArray(
-        chunks=metadata["chunk_grid"]["configuration"]["chunk_shape"],
-        compressor=metadata["codecs"],
+        chunks=chunk_shape,
+        compressor=compressor,
         dtype=np.dtype(metadata["data_type"]),
         fill_value=fill_value,
-        filters=metadata.get("filters", None),
+        filters=filters or None,
         order="C",
-        shape=chunk_shape,
-        zarr_format=3,
+        shape=shape,
+        zarr_format=zarr_format,
     )
 
     return zarray, dim_names, attrs
+
+
+def _configurable_to_num_codec_config(configurable: dict) -> dict:
+    """
+    Convert a zarr v3 configurable into a numcodecs codec.
+    """
+    configurable_copy = configurable.copy()
+    codec_id = configurable_copy.pop("name")
+    configuration = configurable_copy.pop("configuration")
+    return numcodecs.get_codec({"id": codec_id, **configuration}).get_config()
+
+
+def _num_codec_config_to_configurable(num_codec: dict) -> dict:
+    """
+    Convert a numcodecs codec into a zarr v3 configurable.
+    """
+    num_codec_copy = num_codec.copy()
+    return {"name": num_codec_copy.pop("id"), "configuration": num_codec_copy}
+
+
+def _default_fill_value(dtype: np.dtype) -> Union[bool, int, float, str, list]:
```

Comment:
Could this function just be a global variable? Of type `dict[np.dtype: Union[bool, int, float, str, list]]`?

(Also we should be able to use `|` instead of `Union`)

Reply:
Then use `ZARR_DEFAULT_FILLVALUES.get(dtype, default=0.0)`?

Reply:
Oh sorry looks like `.get(dtype, 0.0)` is the right syntax, as `default` is positional-only.